[MSE] Reduce number of decoded frames when a hardware decoder is in use

jyavenard opened the pull request:
#### 43888d71c802d2398073fd30bb4f26fa01223644
<pre>
[MSE] Reduce number of decoded frames when a hardware decoder is in use
<a href="https://bugs.webkit.org/show_bug.cgi?id=291617">https://bugs.webkit.org/show_bug.cgi?id=291617</a>
<a href="https://rdar.apple.com/149362223">rdar://149362223</a>

Reviewed by Youenn Fablet.

The existing logic to limit how many decoded frames we should keep was flawed.
It would decode all compressed frames enqueued and we would re-enqueue compress
frames to always have a maximum of 30. This caused unnecesary elevated
CPU usage and a high number of IOSurfaces.

We now drives the decoding from the purge &amp; display method and stop decoding
when we have 3 (if decoding is HW accelerated) or 10 decoded frames.
The VideoMediaSampleRenderer&apos;s parents are now only driven by the number
of compressed frames already enqueued.

No change in JS observable behaviour. Covered by existing tests

* Source/WebCore/platform/Logging.h: Add MediaPerformance category
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.cpp:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.h: Add link to kVTDecompressionPropertyKey_UsingHardwareAcceleratedVideoDecoder constant.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::areSamplesQueuesReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::maximumDecodedSampleCount const):
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):
(WebCore::VideoMediaSampleRenderer::flushCompressedSampleQueue):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::decodeNextSample): Renamed into decodeNextSampleIfNeeded
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::ensureDecompressionSessionForSample):
(WebCore::WebCoreDecompressionSession::isHardwareAccelerated const):

Canonical link: <a href="https://commits.webkit.org/293860@main">https://commits.webkit.org/293860@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/23e53250b9be1e0c0309b717448c95160d88119e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100109 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19757 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105239 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50690 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102150 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76206 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33274 "Found 60 new test failures: animations/3d/animate-to-transform-perpective-to-none.html animations/3d/change-transform-in-end-event.html animations/3d/full-rotation-animation.html animations/3d/replace-filling-transform.html animations/3d/state-at-end-event-transform.html animations/3d/transform-origin-vs-functions.html animations/3d/transform-perspective.html animations/CSSKeyframesRule-name-null.html animations/CSSKeyframesRule-parameters.html animations/added-while-suspended.html ... (failure)") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103116 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15329 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90413 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56567 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15144 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8410 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50060 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107598 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27222 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19939 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85161 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86619 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84696 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21524 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29370 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7108 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/21075 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27159 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32388 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26970 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30286 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28529 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -440,11 +450,15 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
+#if LOG_DECODER_TIMINGS
+        auto now = MonotonicTime::now();
+        RELEASE_LOG_ERROR(Media, "decoding rate:%0.1fHz decoder rate:%0.1fHz compressed queue:%u decoded queue:%zu", 1.0f / Seconds { now - std::exchange(m_timeSinceLastDecode, now) }.value(), 1.0f / Seconds { now - startTime }.value(), m_compressedSampleQueueSize.load(), decodedSamplesCount());
+#endif
```
Is this needed? If so, I would remove LOG_DECODER_TIMINGS and just use a LOG(Media...).
An alternative is to use `FrameRateMonitor`.

jyavenard replied:
LOG(Media) is only available on debug build AFAIK. And those logs are only of value with release build.
It would also not compile due to `now` variable not being used. 


Those logs are only of value in release build

jyavenard replied:
Added a new MediaPerformance logging category

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -408,6 +411,11 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
+    return decompressionSession.isHardwareAccelerated() ? 3 : 10;
+}
+
 void VideoMediaSampleRenderer::decodeNextSample()
```
decodeNextSampleIfNeeded()?

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -391,14 +393,15 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
     }
 
-    ++m_framesBeingDecoded;
+    m_compressedSampleQueueSize++;
```
Don't we usually ++a or --b instead of c++?

jyavenard replied:
I don't think it's a rule; I personally never use ++a , but I don't mind changing

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -420,6 +428,10 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
+    if (decodedSamplesCount() > maximumDecodedSampleCount(*decompressionSession))
+        return;
+
+    m_compressedSampleQueueSize--;
```
It seems possible that `m_compressedSampleQueueSize` is 0 here. Say if we call `flushCompressedSampleQueue()` in the main thread just before calling `decodeNextSample()` in the dispatcher queue? We would end up with a very big value, not sure whether that has practical impact but it seems best to avoid this.

I would tend to simplify a bit `m_compressedSampleQueueSize` by doing one of these options:
- incrementing/decrementing this value only in dispatcher queue, when appending/taking a sample from `m_compressedSampleQueue`
- Setting `m_compressedSampleQueueSize` to `m_compressedSampleQueue.size()` here if it is important to set aggressively `m_compressedSampleQueueSize` in main thread.

jyavenard replied:
If we do not increase m_compressedSampleQueueSize on the main thread, then `isReadyForMoreMediaData()` will return true until the lambda actually runs on the dispatcher queue ; this can lead to much more data being fed to the VideoMediaSampleRenderer than we want. 

Option 2 seems the most reasonable.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -391,24 +394,30 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
     }
 
-    ++m_framesBeingDecoded;
+    ++m_compressedSampleQueueSize;
```
Do we need this incrementing in main thread, or could we just update `++m_compressedSampleQueueSize` when we modify m_compressedSampleQueue in the lambda?

jyavenard replied:
yes, so that the SourceBufferAVFObjC doesn't potentially append hundreds of frames until m_compressedSampleQueueSize in increased (as until then VideoMediaSampleRenderer::isReadyForMoreMediaData() will return true)

jyavenard committed:
[MSE] Reduce number of decoded frames when a hardware decoder is in use
https://bugs.webkit.org/show_bug.cgi?id=291617
rdar://149362223

Reviewed by Youenn Fablet.

The existing logic to limit how many decoded frames we should keep was flawed.
It would decode all compressed frames enqueued and we would re-enqueue compress
frames to always have a maximum of 30. This caused unnecesary elevated
CPU usage and a high number of IOSurfaces.

We now drives the decoding from the purge & display method and stop decoding
when we have 3 (if decoding is HW accelerated) or 10 decoded frames.
The VideoMediaSampleRenderer's parents are now only driven by the number
of compressed frames already enqueued.

No change in JS observable behaviour. Covered by existing tests

* Source/WebCore/platform/Logging.h: Add MediaPerformance category
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.cpp:
* Source/WebCore/platform/cocoa/VideoToolboxSoftLink.h: Add link to kVTDecompressionPropertyKey_UsingHardwareAcceleratedVideoDecoder constant.
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::areSamplesQueuesReadyForMoreMediaData const):
(WebCore::VideoMediaSampleRenderer::enqueueSample):
(WebCore::VideoMediaSampleRenderer::maximumDecodedSampleCount const):
(WebCore::VideoMediaSampleRenderer::decodeNextSampleIfNeeded):
(WebCore::VideoMediaSampleRenderer::flushCompressedSampleQueue):
(WebCore::VideoMediaSampleRenderer::purgeDecodedSampleQueueAndDisplay):
(WebCore::VideoMediaSampleRenderer::decodeNextSample): Renamed into decodeNextSampleIfNeeded
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.h:
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::WebCoreDecompressionSession::ensureDecompressionSessionForSample):
(WebCore::WebCoreDecompressionSession::isHardwareAccelerated const):

Canonical link: https://commits.webkit.org/293860@main
